lib/utils/config.ts

```diff
@@ -122,9 +109,14 @@ export function generateDeviceConfig(
 			const config = await generateBaseConfig(application, baseConfigOpts);
 
 			if (deviceApiKey == null && semver.satisfies(options.version, '<2.0.3')) {
-				await addApplicationKey(config, application.id);
+				config.apiKey = await sdk.models.application.generateApiKey(
+					application.id,
+				);
 			} else {
-				await addDeviceKey(config, device.uuid, deviceApiKey || true);
+				config.deviceApiKey =
+					typeof deviceApiKey === 'string' && deviceApiKey
+						? deviceApiKey
+						: await sdk.models.device.generateDeviceKey(device.uuid);
```

Comment:
Much better! 👍 
(What was that `deviceApiKey || true` logic?! Good riddance!) 😂

Reply:
Yea right?
Ideally I wanted to do just:
```
config.deviceApiKey = deviceApiKey ?? await sdk.models.device.generateDeviceKey(device.uuid);
```
but that's not 100% how it worked before.
An empty string deviceApiKey is probably bad input though anyway, but I really can't guess whether anyone could do that.
Let me know if you prefer me to include the additional simplification or not.

Reply:
I think the simplification is not possible without additional refactoring. `deviceApiKey` is declared as `deviceApiKey: string | true | undefined` and it may indeed have boolean value `true` to request a device key to be generated, for example in this call: https://github.com/balena-io/balena-cli/blob/v12.52.2/lib/commands/config/generate.ts#L215

I am not a fan of that kind of mixed type API/interface where a value can be a string or a boolean or undefined to trigger different behaviours, but that's how it was implemented and I think it would be low ROI to change it further in this PR.

The PR is good as it is, the important thing was to avoid the key duplication, and as a bonus a bunch of smelly code was deleted and simplified. 👍

Reply:
I think it's fine to do that simplification though since we call it like:
```
config = await generateDeviceConfig(
  device,
  options.deviceApiKey || options['generate-device-api-key'] || undefined,
  answers,
);
```
and then the original implementation of `generateDeviceConfig()` was later doing:
```
await addDeviceKey(config, device.uuid, deviceApiKey || true);
```
which afaict means that undefiend/true/false had no difference for the deviceApiKey param, since the final `if (customDeviceApiKey === true) {` check would give the exact same result in all these cases.

Which leads us to the conclusion that the `--generate-device-api-key` option for `balena config generate` has really no effect...